Combined Frontend for TriFinger, Cameras and Object Tracker

luator opened the pull request:
## Description

This is work in progress, just created a PR to more easily discuss questions.


## How I Tested

TODO


## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [ ] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [ ] All new functions/classes are documented and existing documentation is updated according to changes.
- [ ] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

luator commented on include/robot_fingers/trifinger_platform.hpp:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    // alias
+    robot_interfaces::TriFingerTypes::Observation get_robot_observation(
```
As the class is inheriting from the robot frontend, add this alias which is clearer than just `get_observation`.  Unfortunately, the latter will also still be there, so maybe not the cleanest solution.

luator commented on include/robot_fingers/trifinger_platform.hpp:
```diff
@@ -0,0 +1,90 @@
+        // arrived in between the calls.
+
+
+        // TODO: The implementation below is very naive.
+        // It simply does a linear search starting from the latest time index.
+        // So worst case performance is O(n) where n is the number of object
+        // tracker observations over the period that is covered by the buffer of
+        // the robot data.
+        //
+        // Options to speed this up:
+        // - binary search (?)
+        // - estimate time step size for tracker based on last observations
+        // - store matched indices of last call
+        //
+        // Note, however, that `t` is very likely the latest time index in most
+        // cases.  In this case the match for `t_o` will also be the latest
+        // index of the corresponding time series.  In this case, the complexity
+        // is O(1).  So even when implementing a more complex search algorithm,
+        // the first candidate for `t_o` that is checked should always be the
+        // latest one.
```
I'm not sure if this needs to be improved or if the current naive implementation is maybe already good enough. It depends a bit on how fast the camera will really run in the end (i.e. how big the `n` gets) and how the users are going to use it (i.e. if there is an actual use case to call it with older time indices all the time).

mwuethri commented on include/robot_fingers/trifinger_platform.hpp:
```diff
@@ -0,0 +1,110 @@
+// TODO move implementation to cpp file
+// NOTE: Inheriting from Frontend makes implementation easier but means that we
+// cannot rename get_observation
+class TriFingerPlatform : public robot_interfaces::TriFingerTypes::Frontend
```
ich wuerde wahrscheinlich TriFingerPlatformFrontend nennen :). ist etwas lang, aber faende ich consistenter

mwuethri replied:
i would probably not derive from the Trifinger frontend, but just have a pointer to it, as you do for the tracker.

i am aware that this would lead to a bit more redundant code, but i would find it more logical, and you would also have more control over which functions are available in this class (e.g. i don't know if it makes sense to have a get_observation function in this class, or if we do, it should probably include both, the one from the robot and the one from the tracker) so if deriving at all, i would probably derive from RobotFrontend<Action, Observation> where the Observation includes both..

luator commented:
@mwuethri This is ready for the final review, please take a look again (I refactored a bit and also added the camera frontend).  Not yet tested on the real robot, I'll do this before merging, though.

mwuethri commented:
i will have a look tomorrow

On Mon, Jun 22, 2020 at 4:54 PM Felix Widmaier <notifications@github.com>
wrote:

> @mwuethri <https://github.com/mwuethri> This is ready for the final
> review, please take a look again (I refactored a bit and also added the
> camera frontend). Not yet tested on the real robot, I'll do this before
> merging, though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-dynamic-robot-initiative/robot_fingers/pull/27#issuecomment-647569731>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAO6EOWYWF7HXFMBSBG7QFDRX5WCFANCNFSM4OBLG4WQ>
> .
>
